wallet-new/src/Cardano/Wallet/API/V1/Migration.hs

```diff
@@ -0,0 +1,105 @@
+{- | This is a temporary module to help migration @V0@ datatypes into @V1@ datatypes.
+-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE TypeSynonymInstances  #-}
+module Cardano.Wallet.API.V1.Migration (
+      MonadV1
+    , V1Context
+    , v1MonadNat
+    , lowerV1Monad
+    , Migrate(..)
+    -- * Isomorphisms
+    , walletAssurance
+    -- * Configuration re-exports
+    , HasInfraConfiguration
+    , HasSscConfiguration
+    , HasCompileInfo
+    , HasConfiguration
+    , HasUpdateConfiguration
+    , HasNodeConfiguration
+    ) where
+
+import           Universum
+
+import qualified Cardano.Wallet.API.V1.Types      as V1
+import qualified Pos.Wallet.Web.ClientTypes.Types as V0
+
+import           Control.Lens
+import qualified Control.Monad.Catch              as Catch
+import           Mockable                         (runProduction)
+import           Servant
+
+import           Pos.Configuration                (HasNodeConfiguration)
+import           Pos.Core.Configuration           (HasConfiguration)
+import           Pos.Infra.Configuration          (HasInfraConfiguration)
+import           Pos.Ssc                          (HasSscConfiguration)
+import           Pos.Update.Configuration         (HasUpdateConfiguration)
+import           Pos.Util.CompileInfo             (HasCompileInfo)
+import           Pos.Wallet.Web.Mode              (WalletWebMode, WalletWebModeContext)
+
+-- | Temporary monad to handle the migration from the V0 & V1 stacks.
+type MonadV1   = WalletWebMode
+type V1Context = WalletWebModeContext
+
+-- | Hoist a 'V1' monad to a Servant's Handler.
+-- See: http://haskell-servant.readthedocs.io/en/stable/tutorial/Server.html#natural-transformations
+v1MonadNat :: V1Context -> (forall a. MonadV1 a -> Handler a)
+v1MonadNat = lowerV1Monad
+
+-- | Converts our domain-specific monad into a standard Servant `Handler`.
+lowerV1Monad :: V1Context -> MonadV1 a -> Handler a
+lowerV1Monad ctx handler =
+    liftIO (hoistHandler handler) `Catch.catches` excHandlers
+  where
+
+    hoistHandler :: forall a . MonadV1 a -> IO a
+    hoistHandler = runProduction . flip runReaderT ctx
+
+    excHandlers = [Catch.Handler throwError]
+
+-- | 'Migrate' encapsulates migration between types, when possible.
+class Migrate from to where
+    migrate :: from -> to
+
+--
+-- Instances
+--
+
+instance Migrate V0.CWallet V1.Wallet where
+    migrate V0.CWallet{..} =
+        V1.Wallet { V1.walId   = migrate @(V0.CId V0.Wal) cwId
+                  , V1.walName = V0.cwName cwMeta
+                  , V1.walBalance = migrate @V0.CCoin @V1.RenderedBalance cwAmount
```

Comment:
Just wondering: doesn't GHC  deduce these types automatically given type of `cwAmount` and `walBalance`? I believe it would be less verbose to not write these type applications, if it does.

You will probably say that explicit type applications help to understand quicker which exact `Migrate` instances are used here, but I don't think that this effect is significant (though it's arguable). It's minor anyway.

Reply:
You are probably right, GHC should be smart enough and as far as we are concerned this module will be removed eventually, so I'm happy either way. I will try to remove the type annotation, and we will re-assess later on if that compromised readability too much.